agora/Agora/Governor/Scripts.hs

```diff
@@ -0,0 +1,776 @@
+      * At least 'Agora.Stake.stackedAmount' of GT must be spent in the transaction.
+      * The transaction must be signed by the stake owner.
+
+  - Exactly one new proposal state token is minted.
+  - An UTXO which holds the newly minted proposal state token is sent to the proposal validator.
+    This UTXO must have a valid datum of type 'Agora.Proposal.ProposalDatum', the datum must:
+
+      * Copy its id and thresholds from the governor's state.
+      * Have status set to 'Proposal.Draft'.
+      * Have zero votes.
+      * Have exactly one cosigner - the stake owner
+
+  - An UTXO which holds the stake state token is sent back to the stake validator.
+    This UTXO must have a valid datum of type 'Agora.Stake.StakeDatum':
+
+      * The 'Agora.Stake.stakedAmount' and 'Agora.Stake.owner' should not be changed,
+         comparing to the input stake.
+      * The new proposal locks must be appended to the 'Agora.Stake.lockedBy'.
+
+  == Minting GATs
+
+  When the redeemer is 'MintGATs', the script will check:
+
+  - Governor's state is not changed.
+  - Exactly only one proposal is in the inputs. Let's call this the /input proposal/.
+  - The proposal is in the 'Proposal.Executable' state.
+
+  NOTE: The input proposal is found by looking for the UTXO with a proposal state token in the inputs.
+
+  === Effect Group Selection
+
+  Currently a proposal can have two or more than two options to vote on,
+    meaning that it can contains two or more effect groups,
+    according to [#39](https://github.com/Liqwid-Labs/agora/issues/39).
+
+  Either way, the shapes of 'Proposal.votes' and 'Proposal.effects' should be the same.
+    This is checked by 'Proposal.proposalDatumValid'.
+
+  The script will look at the the 'Proposal.votes' to determine which group has the highest votes,
+    said group shoud be executed.
+
+  During the process, minimum votes requirement will also be enforced.
+
+  Next, the script will:
+
+  - Ensure that for every effect in the said effect group,
+    exactly one valid GAT is minted and sent to the effect.
+  - The amount of GAT minted in the transaction should be equal to the number of effects.
+  - A new UTXO is sent to the proposal validator, this UTXO should:
+
+      * Include the one proposal state token.
+      * Have a valid datum of type 'Proposal.ProposalDatum'.
+        This datum should be as same as the one of the input proposal,
+        except its status should be 'Proposal.Finished'.
+
+  == Changing the State
+
+  Redeemer 'MutateGovernor' allows the state datum to be changed by an external effect.
+
+  In this case, the script will check
+
+  - Exactly one GAT is burnt in the transaction.
+  - Said GAT is tagged by the effect.
+-}
+governorValidator :: Governor -> ClosedTerm PValidator
+governorValidator gov =
+  plam $ \datum' redeemer' ctx' -> P.do
+    (pfromData -> redeemer, _) <- ptryFrom redeemer'
+    ctx <- pletFields @'["txInfo", "purpose"] ctx'
```

Comment:
Could you add some comments through the body of the do-block to help guide the reader through what you're doing at each step?